modules/ssl/ssl_engine_init.c

```diff
@@ -903,7 +1022,39 @@ static apr_status_t ssl_init_ctx_protocol(server_rec *s,
         SSL_CTX_set_options(ctx, SSL_OP_IGNORE_UNEXPECTED_EOF);
     }
 #endif
-    
+
+#ifdef HAVE_OPENSSL_ECH
+#if SSL_HAVE_PROTOCOL_TLSV1_3
+
+    /* ECH only really makes sense for TLSv1.3 */
+    prot = SSL_CTX_get_max_proto_version(ctx);
+    if (sc->echkeydir) {
+        if (prot == TLS1_3_VERSION) {
+            /* try load the keys */
+            if (load_echkeys(ctx,sc->echkeydir,s,ptemp) != 0) {
+                ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(10531)
+                    "ECHKeyDir failed to load keys - exiting");
+                SSL_CTX_free(ctx);
+                mctx->ssl_ctx = NULL;
+                return ssl_die(s);
+            }
+        } else {
+            ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(10532)
+                 "ECHKeyDir configured but TLSv1.3 turned off - exiting.");
+            SSL_CTX_free(ctx);
+            mctx->ssl_ctx = NULL;
+            return ssl_die(s);
+        } 
+    }
+
+#else
+    if (sc->echkeydir) {
+        ap_log_error(APLOG_MARK, APLOG_WARN, 0, s, APLOGNO(10533)
+                 "ECHKeyDir configured but no TLSv1.3 so ECH will be ignored.");
```

Comment:
IMO if TLSv1.3 is not supported by the TLS toolkit, using `SSLECHKeyDir` should be a config-time error, so merely warning for this is not sufficient. So in the definition of `HAVE_OPENSSL_ECH` in ssl_private.h check that SSL_HAVE_PROTOCOL_TLSV1_3 definition too, and drop this case?

Reply:
Having thought about that, I figure the better thing to do is number 2 above - treat all these errors as fatal and don't start the server (so use `APLOG_EMERG`) - what convinced me was that if you start an old server that doesn't have any ECH support, and that sees an `SSLECHKeyDir` directive then that old server will fail to start, so doing likewise in these cases seems correct. So I think the code now in the PR does the right thing.

@notroj - is there a reason to do this in `ssl_cmd_SSLECHKeyDir` or is it ok as-is?

Reply:
Just pushed a change for that, let me know if more needed - the commit is [here](https://github.com/apache/httpd/pull/551/commits/5a4db6c1b0412260ea28435ccf34664a1ccb5ef3)